[inductor] Fix cudagraph skip for index_put_ with boolean indices, graph partitioning logic

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #170103

A few fixes:

- we weren't partitioning around index put with boolean inputs

- graph partitioning was skipping the whole graph, anytime we used V.graph.disable_cudagraphs_reason. There's no reason to use this for partitioning. I've updated the skip logic to encompass all of the reasons we would use disable cudagraphs reason.

- Prune the deterministic disable cudagraphs reason. I'm not sure how this list of ops got there originally, but I've added opinfo tests showing they're cudagraphable.

We run into errors with just part of these fixes, unrelated, so i'm doing this as one pr.



Fixes #169951

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

BoyuanFeng commented on test/inductor/test_cudagraph_trees.py:
```diff
@@ -4682,6 +4716,77 @@ def fn(x, y):
+                    # Run multiple times to trigger cudagraph recording
+                    for _ in range(3):
+                        input_clone = sample.input.clone()
+                        result = compiled_fn(input_clone, *sample.args, **sample.kwargs)
```
Should we also assert on the number of graph partition / cudagraph skips?

eellison commented on torch/_inductor/utils.py:
```diff
@@ -1162,57 +1162,33 @@ def any_is_symbolic(*args: Any) -> bool:
-                "aten.scatter_add.default",
-                "aten.scatter_reduce.two",
-                "aten.scatter_reduce_.two",
-                "aten.scatter_reduce.two_out",
```
Yes, I added a test that shows they're all cudagraph able. I'm not sure where this list even comes from,

ngimel replied:
It comes from broad index_put ban, under deterministic mode `scatter` dispatches to `index_put`, but with `Long` indices so that's fine.

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update on "[inductor] Fix cudagraph skip for index_put_ with boolean indices, graph partitioning logic"


A few fixes:

- we weren't partitioning around index put with boolean inputs

- graph partitioning was skipping the whole graph, anytime we used V.graph.disable_cudagraphs_reason. There's no reason to use this for partitioning. I've updated the skip logic to encompass all of the reasons we would use disable cudagraphs reason.

- Prune the deterministic disable cudagraphs reason. I'm not sure how this list of ops got there originally, but I've added opinfo tests showing they're cudagraphable.

We run into errors with just part of these fixes, unrelated, so i'm doing this as one pr.



Fixes #169951

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

eellison committed:
Update on "[inductor] Fix cudagraph skip for index_put_ with boolean indices, graph partitioning logic"


A few fixes:

- we weren't partitioning around index put with boolean inputs

- graph partitioning was skipping the whole graph, anytime we used V.graph.disable_cudagraphs_reason. There's no reason to use this for partitioning. I've updated the skip logic to encompass all of the reasons we would use disable cudagraphs reason.

- Prune the deterministic disable cudagraphs reason. I'm not sure how this list of ops got there originally, but I've added opinfo tests showing they're cudagraphable.

We run into errors with just part of these fixes, unrelated, so i'm doing this as one pr.



Fixes #169951

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update on "[inductor] Fix cudagraph skip for index_put_ with boolean indices, graph partitioning logic"


A few fixes:

- we weren't partitioning around index put with boolean inputs

- graph partitioning was skipping the whole graph, anytime we used V.graph.disable_cudagraphs_reason. There's no reason to use this for partitioning. I've updated the skip logic to encompass all of the reasons we would use disable cudagraphs reason.

- Prune the deterministic disable cudagraphs reason. I'm not sure how this list of ops got there originally, but I've added opinfo tests showing they're cudagraphable.

We run into errors with just part of these fixes, unrelated, so i'm doing this as one pr.



Fixes #169951

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

eellison committed:
Update on "[inductor] Fix cudagraph skip for index_put_ with boolean indices, graph partitioning logic"


A few fixes:

- we weren't partitioning around index put with boolean inputs

- graph partitioning was skipping the whole graph, anytime we used V.graph.disable_cudagraphs_reason. There's no reason to use this for partitioning. I've updated the skip logic to encompass all of the reasons we would use disable cudagraphs reason.

- Prune the deterministic disable cudagraphs reason. I'm not sure how this list of ops got there originally, but I've added opinfo tests showing they're cudagraphable.

We run into errors with just part of these fixes, unrelated, so i'm doing this as one pr.



Fixes #169951

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

eellison committed:
Update on "[inductor] Fix cudagraph skip for index_put_ with boolean indices, graph partitioning logic"


A few fixes:

- we weren't partitioning around index put with boolean inputs

- graph partitioning was skipping the whole graph, anytime we used V.graph.disable_cudagraphs_reason. There's no reason to use this for partitioning. I've updated the skip logic to encompass all of the reasons we would use disable cudagraphs reason.

- Prune the deterministic disable cudagraphs reason. I'm not sure how this list of ops got there originally, but I've added opinfo tests showing they're cudagraphable.

We run into errors with just part of these fixes, unrelated, so i'm doing this as one pr.



Fixes #169951

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

izaitsevfb commented:
@pytorchbot revert -c ghfirst -m "breaks inductor:cudagraph_trees_expandable_segments tests internally, see [D89125296](https://www.internalfb.com/diff/D89125296)"

eellison committed:
Update on "[inductor] Fix cudagraph skip for index_put_ with boolean indices, graph partitioning logic"


A few fixes:

- we weren't partitioning around index put with boolean inputs

- graph partitioning was skipping the whole graph, anytime we used V.graph.disable_cudagraphs_reason. There's no reason to use this for partitioning. I've updated the skip logic to encompass all of the reasons we would use disable cudagraphs reason.

- Prune the deterministic disable cudagraphs reason. I'm not sure how this list of ops got there originally, but I've added opinfo tests showing they're cudagraphable.

We run into errors with just part of these fixes, unrelated, so i'm doing this as one pr.



Fixes #169951

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela

[ghstack-poisoned]

eellison commented:
@pytorchbot merge -i

eellison committed:
Update

[ghstack-poisoned]

eellison commented:
@pytochbot merge

eellison commented:
@pytorchbot merge

eellison commented:
@pytorchbot merge -i

eellison committed:
Update

[ghstack-poisoned]

Lucaskabela commented:
@pytorchbot merge -i

Lucaskabela commented:
@pytorchbot cherry-pick --onto release/2.10 --fixes "Fixes https://github.com/pytorch/pytorch/issues/169951"  -c regressions

Lucaskabela commented:
@pytorchbot cherry-pick --onto release/2.10 --fixes "Fixes https://github.com/pytorch/pytorch/issues/169951" -c regression
